util: ParseByteUnits - Parse a string with suffix unit

dougEfresh opened the pull request:
A convenience utility for parsing human readable strings sizes e.g. `500G` is `500 * 1 << 30`

The argument/setting `maxuploadtarget`  now accept human readable byte units `[k|K|m|M|g|G||t|T]`
This change  backward compatible, defaults to `M` if no unit specified.

promag reviewed (commented):
Concept ACK. However, I think the implementation can be improved. Why not check the last char first?  If it's not a digit then extract it and validate to get the multiplier. Only then parse the number.

dougEfresh commented on src/init.cpp:
```diff
@@ -1749,7 +1750,11 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
-    connOptions.nMaxOutboundLimit = 1024 * 1024 * args.GetIntArg("-maxuploadtarget", DEFAULT_MAX_UPLOAD_TARGET);
+    auto opt_max_upload = ParseByteUnits(args.GetArg("-maxuploadtarget", DEFAULT_MAX_UPLOAD_TARGET));
+    if (!opt_max_upload) {
```
I want to point out to reviewers that this changes start up behavior. Previously,  `-maxuploadtarget`  would silently fail and use DEFAULT_MAX_UPLOAD_TARGET. 
This change will throw an error upon parse failure, but only  after 30-50 seconds after start-up

dougEfresh replied:
Moved

ryanofsky commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,63 @@ std::string HexStr(const Span<const uint8_t> s)
+        return opt_uint64{nBytes};
+    }
+    // check overflow
+    return nBytes > std::numeric_limits<uint64_t>::max() / multiplier ? std::nullopt : opt_uint64{nBytes * multiplier};
```
In commit "util: ParseByteUnits - Parse a string with suffix unit [k|K] [m|M] [g|G] [t|T]" (c5d4a59cf91f54a2027851b32cf83cfcb04cd210)

There's repetition and unneeded variables in this function. Also `nBytes` holds the number before multiplication instead of the actual number of bytes, and the name doesn't follow current coding convention. Would suggest an overall cleanup here like:

```diff
diff --git a/src/util/strencodings.cpp b/src/util/strencodings.cpp
index c490cc7f3e1..0075e69c5e7 100644
--- a/src/util/strencodings.cpp
+++ b/src/util/strencodings.cpp
@@ -530,19 +530,8 @@ std::string HexStr(const Span<const uint8_t> s)
 
 std::optional<uint64_t> ParseByteUnits(const std::string& str, const uint64_t default_multiplier)
 {
-    if (str.empty()) {
-        return std::nullopt;
-    }
-
-    using opt_uint64 = std::optional<uint64_t>;
-
-    uint64_t nBytes{0};
-    uint64_t multiplier{default_multiplier};
-    auto unit = str.back();
-    auto v = str;
-
-    v.pop_back();
-
+    auto multiplier = default_multiplier ? default_multiplier : 1;
+    char unit = str.empty() ? 0 : str.back();
     switch (unit) {
         case 'k':
             multiplier = 1000;
@@ -569,21 +558,13 @@ std::optional<uint64_t> ParseByteUnits(const std::string& str, const uint64_t de
             multiplier = 1ULL << 40;
             break;
         default:
-            // unit should be digit, parse original str
-            if (!ParseUInt64(str, &nBytes)) {
-                return std::nullopt;
-            }
-            //check overflow and use default_multiplier
-            return default_multiplier == 0 ? opt_uint64{nBytes} : nBytes > std::numeric_limits<uint64_t>::max() / default_multiplier ? std::nullopt : opt_uint64{nBytes * default_multiplier};
+            unit = 0;
+            break;
     }
-
-    if (!ParseUInt64(v, &nBytes)) {
+    uint64_t num;
+    if (!ParseUInt64(unit ? str.substr(0, str.size()-1) : str, &num) ||
+        num > std::numeric_limits<uint64_t>::max() / multiplier) {
         return std::nullopt;
     }
-
-    if (multiplier == 0) {
-        return opt_uint64{nBytes};
-    }
-    // check overflow
-    return nBytes > std::numeric_limits<uint64_t>::max() / multiplier ? std::nullopt : opt_uint64{nBytes * multiplier};
+    return num * multiplier;
 }

```

dougEfresh replied:
Great suggestion, code now looks cleaner.

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -2456,4 +2456,45 @@ BOOST_AUTO_TEST_CASE(remove_prefix)
+    BOOST_CHECK_EQUAL(ParseByteUnits("4T").value_or(0), 4ULL << 40);
+
+    // override default multiplier
+    BOOST_CHECK_EQUAL(ParseByteUnits("5", 1ULL << 10).value_or(0), 5ULL << 10);
```
In commit "util: ParseByteUnits - Parse a string with suffix unit [k|K] [m|M] [g|G] [t|T]" (c5d4a59cf91f54a2027851b32cf83cfcb04cd210)

Would be good to check behavior with 0 multiplier to make sure it doesn't accidentally change in the future.

dougEfresh replied:
Added `0` multiplier test

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -2456,4 +2456,45 @@ BOOST_AUTO_TEST_CASE(remove_prefix)
+    BOOST_CHECK(!ParseByteUnits("-123m").has_value());
+
+    // zero padding
+    BOOST_CHECK_EQUAL(ParseByteUnits("020M").value_or(1), 20ULL << 20);
```
In commit "util: ParseByteUnits - Parse a string with suffix unit [k|K] [m|M] [g|G] [t|T]" (c5d4a59cf91f54a2027851b32cf83cfcb04cd210)

Minor inconsistency here where this case is using `.value_or(1)` and cases above are using `.value_or(0)`. I think it would best if all of these cases were consistent and better if they were using `.value()` instead of `.value_or()`

dougEfresh replied:
Agreed

ryanofsky reviewed (approved):
Code review and lightly tested ACK c5d4a59cf91f54a2027851b32cf83cfcb04cd210. I left some suggestions below, but none are important so feel free to ignore them.

It might be good to add a release note in the `doc/` directory mentioning the new argument verification, since in theory it could make a configuration that used to work previously now cause an error on startup.

It would also be good to mention new behavior in the PR description, since current description is a little ambiguous about whether this is only adding a utility function or changing behavior.

dougEfresh commented:
> It might be good to add a release note in the `doc/` directory mentioning the new argument verification, since in theory it could make a configuration that used to work previously now cause an error on startup.

Added to release docs

> It would also be good to mention new behavior in the PR description, since current description is a little ambiguous about whether this is only adding a utility function or changing behavior.

Updated PR description. 

Thanks for feedback @ryanofsky

ryanofsky reviewed (approved):
Code review ACK f1f8a772b0daf8cc0c40f0fc54e82618ad242ddd. Just suggested changes since last review (thanks!)

vasild commented on src/init.cpp:
```diff
@@ -1748,8 +1755,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     connOptions.m_added_nodes = args.GetArgs("-addnode");
-
-    connOptions.nMaxOutboundLimit = 1024 * 1024 * args.GetIntArg("-maxuploadtarget", DEFAULT_MAX_UPLOAD_TARGET);
+    connOptions.nMaxOutboundLimit = *opt_max_upload;;
```
nit: double `;;` at the end

dougEfresh replied:
removed

vasild commented on src/test/util_tests.cpp:
```diff
@@ -2456,4 +2456,49 @@ BOOST_AUTO_TEST_CASE(remove_prefix)
     BOOST_CHECK_EQUAL(RemovePrefix("", ""), "");
 }
 
+BOOST_AUTO_TEST_CASE(util_ParseByteUnits)
```
Maybe add a test for unknown suffix, e.g. `123X`.

dougEfresh replied:
Added invalid unit test

vasild commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+    }
+
+    uint64_t num;
+    if (!ParseUInt64(unit ? str.substr(0, str.size()-1) : str, &num) ||
```
nit: missing spaces around `-`.

dougEfresh replied:
When I added the space, it didn't look aesthetically pleasing to me, normally I do add a space, but this if statement looks better without 
I don't mind adding a space if you like or if someone else asks.

vasild replied:
The thing is that lots of people have lots of different preferences. "Aesthetically pleasant" for somebody may not be for another. The project uses `clang-format` to avoid endless low-value discussions about styling.

dougEfresh replied:
@vasild I changed `const uint64_t`  to `uint64_t`

vasild commented on src/init.cpp:
```diff
@@ -435,7 +436,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-maxsendbuffer=<n>", strprintf("Maximum per-connection send buffer, <n>*1000 bytes (default: %u)", DEFAULT_MAXSENDBUFFER), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-maxtimeadjustment", strprintf("Maximum allowed median peer time offset adjustment. Local perspective of time may be influenced by peers forward or backward by this amount. (default: %u seconds)", DEFAULT_MAX_TIME_ADJUSTMENT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-maxuploadtarget=<n>", strprintf("Tries to keep outbound traffic under the given target (in MiB per 24h). Limit does not apply to peers with 'download' permission. 0 = no limit (default: %d)", DEFAULT_MAX_UPLOAD_TARGET), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-maxuploadtarget=<n>", strprintf("Tries to keep outbound traffic under the given target (in MiB, GiB or TiB per 24h). Limit does not apply to peers with 'download' permission or blocks created within past week. 0 = no limit (default: %s). Optional suffix units [m|M] MiB, [g|G] GiB, [t|T] TiB. Lowercase is 1000 base while uppercase is 1024 base (default).", DEFAULT_MAX_UPLOAD_T ...
```
I find this text confusing:

> Optional suffix units [m|M] MiB, [g|G] GiB, [t|T] TiB. Lowercase is 1000 base while uppercase is 1024 base

Maybe:

> Optional suffix units [m|M|g|G|t|T]. Lowercase is 1000 base while uppercase is 1024 base

dougEfresh replied:
Good suggestion, I've changed the doc to: 
```
Tries to keep outbound traffic under the given target per 24h. Limit does not apply to peers with 'download' permission or blocks created within past week. 0 = no limit (default: %s). Optional suffix units [k|K|m|M|g|G|t|T] (default: M). Lowercase is 1000 base while uppercase is 1024 base
```

vasild commented on src/util/strencodings.h:
```diff
@@ -305,4 +305,17 @@ std::string ToUpper(const std::string& str);
+ * @returns                        optional uint64_t bytes from str or nullopt
+ *                                 if ParseUInt64 is false, str is empty, trailing whitespace or overflow
+ */
+std::optional<uint64_t> ParseByteUnits(const std::string& str, const uint64_t default_multiplier = 1024UL*1024UL);
```
nit: drop unnecessary `const` from the integer argument
style: missing spaces around `*`:

```suggestion
std::optional<uint64_t> ParseByteUnits(const std::string& str, uint64_t default_multiplier = 1024UL * 1024UL);
```

vasild replied:
The `const` is still there. I think there was discussion somewhere about it, but I can't find it now...

vasild commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+    uint64_t num;
+    if (!ParseUInt64(unit ? str.substr(0, str.size()-1) : str, &num) ||
+        num > std::numeric_limits<uint64_t>::max() / multiplier) { // check overflow
+         return std::nullopt;
```
style: this has 5 spaces for indentation, should be 4.

vasild commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+{
+    auto multiplier = default_multiplier ? default_multiplier : 1;
+    char unit = str.empty() ? 0 : str.back();
+    switch (unit) {
+        case 'k':
```
style: `case` should be aligned below `switch`:

```cpp
    switch (unit) {
    case 'k':
        multiplier = 1000;
```

vasild commented on src/init.cpp:
```diff
@@ -1114,6 +1115,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+    auto opt_max_upload = ParseByteUnits(args.GetArg("-maxuploadtarget", DEFAULT_MAX_UPLOAD_TARGET));
+    if (!opt_max_upload) {
+        return InitError(strprintf(_("Unable to parse -maxuploadtarget: '%s' (possible integer overflow?)"), args.GetArg("-maxuploadtarget","")));
```
style: missing space

```suggestion
        return InitError(strprintf(_("Unable to parse -maxuploadtarget: '%s' (possible integer overflow?)"), args.GetArg("-maxuploadtarget", "")));
```

vasild reviewed (approved):
ACK f3f88062f4cdb0cc2d1af221a5a9da9ee2ea0e46

Some styling nits below, which would be good to address before merge.

The style rules are defined in `src/.clang-format`. You can use `contrib/devtools/clang-format-diff.py` to properly format the patch.

vasild reviewed (approved):
ACK 6ebb47626ee30576548e8b3c01ad961761e28562

dougEfresh commented:
@MarcoFalke  would you mind reviewing? Thanks

vasild reviewed (approved):
ACK 4451a0000dda5daba4b1624dce1c95f1914e7251

promag commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+
+std::optional<uint64_t> ParseByteUnits(const std::string& str, uint64_t default_multiplier)
+{
+    auto multiplier = default_multiplier ? default_multiplier : 1;
```
nit, why not `assert(default_multiplier > 0)`?

dougEfresh replied:
Depends on how strict or how we want to use this function.
For me, a `0` default_multiplier means  "don not apply a multiplier"  if no unit found, parse  `str` as is. This may not be intuitive. 
I'm fine with restricting multiplier > 0.  
@MarcoFalke what's your opinion ? 
either way, I'll update the docs in strencoding.h to explaining what a default_multiplier of zero means.

MarcoFalke replied:
Wouldn't "don not apply a multiplier" mean `1`?

dougEfresh replied:
ok, i will add an assert

promag commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+std::optional<uint64_t> ParseByteUnits(const std::string& str, uint64_t default_multiplier)
+{
+    auto multiplier = default_multiplier ? default_multiplier : 1;
+    char unit = str.empty() ? 0 : str.back();
```
nit,
```cpp
if (str.empty()) return std::nullopt;
char unit = str.back();
```

promag reviewed (commented):
Code review ACK 4451a0000dda5daba4b1624dce1c95f1914e7251.

MarcoFalke commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,45 @@ std::string HexStr(const Span<const uint8_t> s)
+    }
+
+    uint64_t num;
+    if (!ParseUInt64(unit ? str.substr(0, str.size() - 1) : str, &num) ||
```
With this breaking change (no longer ignoring whitespace), which I think is fine, it might be best to also reject sign characters (+-) and use `ToIntegral<uint64_t>`.

dougEfresh replied:
I have tests for +-, whitespace, invalid unit...etc
I'm now using ToIntegral. 
I've added an `assert(default_multiplier > 0)` but I don't know how to test this. Any ideas examples?

vasild replied:
You cannot have the test trigger an `assert()` and continue successfully. Instead, you can `throw std::out_of_range` from the function and `try/catch` it from the test (`BOOST_CHECK_THROW()`). But in this case maybe it is best/simpler to just remove the parameter. Then there is no need to test edge cases for it :)

MarcoFalke commented on src/util/strencodings.h:
```diff
@@ -305,4 +305,17 @@ std::string ToUpper(const std::string& str);
+ * @returns                        optional uint64_t bytes from str or nullopt
+ *                                 if ParseUInt64 is false, str is empty, trailing whitespace or overflow
+ */
+std::optional<uint64_t> ParseByteUnits(const std::string& str, uint64_t default_multiplier = 1024UL * 1024UL);
```
Should the default multiplier have a default? It seems like the default is something that is per-arg and not project-wide.

dougEfresh replied:
Agreed.  I remove the default.

dougEfresh replied:
How about `1` as a default multiplier ?

promag replied:
How about removing the argument? It's not needed. Either way please update doc comment above.

dougEfresh replied:
default of `1` seems natural, so my latest change defaults to `1`

vasild replied:
> How about removing the argument? It's not needed.

Indeed! I would remove that unused code. No parameter, no problem :) (what should be its default value? should it have default value? how to test `0` if we have assert()?)

dougEfresh replied:
i do need this multiplier argument. the intent of this function is to be used with maxupload, dbcache...etc. they inheritly have a default multiplier (mostly MiB).  the main purpose is maintain backwards capability, e. g. `maxupload=300`  becomes `300m`.
As I write this, i think I will change the argument to a an enum with the valid units [k|K|m|M...]
with no default value. This makes it clear and easy to verify/test.

dougEfresh replied:
@vasild @MarcoFalke @promag 
I added an enum `ByteUnitMultiplier` to [src/util/strencodings.src/util/strencodings.h](https://github.com/bitcoin/bitcoin/pull/23249/files#r748048958) 
Let me know what you think.

MarcoFalke commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,46 @@ std::string HexStr(const Span<const uint8_t> s)
+        multiplier = 1ULL << 20;
+        break;
+    case 'g':
+        multiplier = 1000ULL * 1000 * 1000;
```
```suggestion
        multiplier = 1000'000'000ULL;
```

Haven't tried this, but can the compiler parse this?

dougEfresh replied:
it does!

MarcoFalke commented on src/util/strencodings.cpp:
```diff
@@ -526,3 +527,46 @@ std::string HexStr(const Span<const uint8_t> s)
+
+    auto parsed_num = ToIntegral<uint64_t>(unit ? str.substr(0, str.size() - 1) : str);
+    // check overflow
+    return !parsed_num || parsed_num > std::numeric_limits<uint64_t>::max() / multiplier ? std::nullopt : std::optional{*parsed_num * multiplier};
```
```suggestion
    if( !parsed_num || parsed_num > std::numeric_limits<uint64_t>::max() / multiplier){
    return std::nullopt;
    }
    return *parsed_num * multiplier;
```

Maybe split up into two lines to avoid the `std::optional{}` constructor for the value?

dougEfresh replied:
I agree, it looks better split

maflcko reviewed (commented):
feel free to ignore the style nits

dougEfresh commented on src/util/strencodings.h:
```diff
@@ -29,6 +29,22 @@ enum SafeChars
+    G_IEC = 1ULL << 30,
+    T_SI = 1'000'000'000'000ULL,
+    T_IEC = 1ULL << 40,
+};
```
I use the "technical" terms but I don't think they are intuitive, open to suggestions.

promag commented on src/util/strencodings.h:
```diff
@@ -29,6 +29,22 @@ enum SafeChars
+ * Used by ParseByteUnits()
+ * SI are 1000 base while IEC is 1024 base
+ */
+enum ByteUnitMultiplier {
```
Use enum class, which can be used like `ByteUnit::g` or `ByteUnit::K`.

dougEfresh replied:
changed. more intuitive, clear.

dougEfresh commented:
Windows doesn't seem to like uint64_t enums: 
```
src\util\strencodings.h(37,21): error C3064: '__uint64_t': must be a simple type or resolve to one
enumerator value '1099511627776' cannot be represented as 'int', value is '0'
```
Suggestions ?

ryanofsky reviewed (approved):
Code review ACK 1f51a25aa5037210ba39e14f048b3c9ec1d342ca. Main changes since last review are removing default multiplier, asserting multiplier is positive, adding enum

vasild commented on src/test/util_tests.cpp:
```diff
@@ -2456,4 +2456,52 @@ BOOST_AUTO_TEST_CASE(remove_prefix)
+    BOOST_CHECK(!ParseByteUnits("0.5T", noop));
+
+    // overflow
+    BOOST_CHECK(!ParseByteUnits("18446744073709551615g", noop).has_value());
```
nit: remove `.has_value()` for consistency with other tests:

```suggestion
    BOOST_CHECK(!ParseByteUnits("18446744073709551615g", noop));
```

dougEfresh replied:
fixed in 21b58f430fa05fdb7c5db79b545302417a5dbceb

vasild commented on src/util/strencodings.h:
```diff
@@ -305,4 +322,17 @@ std::string ToUpper(const std::string& str);
 std::string Capitalize(std::string str);
 
+/**
+ * Parse a string with suffix unit [k|K|m|M|g|G|t|T] .
```
nit: extra white space

```suggestion
 * Parse a string with suffix unit [k|K|m|M|g|G|t|T].
```

dougEfresh replied:
fixed

vasild reviewed (commented):
The code looks ok (1f51a25aa5037210ba39e14f048b3c9ec1d342ca). Some nits below, feel free to ignore.

I think the 3 commits should be squashed into one - subsequent ones just change stuff added by the first commit.

dougEfresh committed:
util: ParseByteUnits - Parse a string with suffix unit [k|K|m|M|g|G|t|T]

A convenience utility for human readable arguments/config e.g. -maxuploadtarget=500g

dougEfresh commented:
@MarcoFalke 
hopefully, last review [21b58f43](https://github.com/bitcoin/bitcoin/commit/21b58f430fa05fdb7c5db79b545302417a5dbceb)
Main changes since last review are removing assigning a default multiplier, adding enum for ByteUnit(s)

vasild reviewed (approved):
ACK 21b58f430fa05fdb7c5db79b545302417a5dbceb

Just squash and 2 minor nits since 1f51a25aa5, verified with
```sh
diff -u <(git diff 1f51a25aa5~3..1f51a25aa5) <(git show 21b58f430f) |vim -
```

Thanks!

ryanofsky reviewed (approved):
Code review ACK 21b58f430fa05fdb7c5db79b545302417a5dbceb. Only changes since last review are dropping optional has_value call, fixing comment punctuation, squashing commits.

MarcoFalke commented on src/test/util_tests.cpp:
```diff
@@ -2456,4 +2456,52 @@ BOOST_AUTO_TEST_CASE(remove_prefix)
+    BOOST_CHECK_EQUAL(ParseByteUnits("1", noop).value(), 1);
+    BOOST_CHECK_EQUAL(ParseByteUnits("0", noop).value(), 0);
+
+    BOOST_CHECK_EQUAL(ParseByteUnits("1k", noop).value(), 1000ULL);
```
nit: Could also add a test that default_multiplier is ignored completely if passed in via the string? Maybe also add a fuzz test?

rebroad commented:
This fails to compile on some (unsupported, e.g. cygwin) systems:

```
  CXX      bitcoind-bitcoind.o
In file included from ./netaddress.h:19,
                 from ./chainparams.h:11,
                 from bitcoind.cpp:10:
./util/strencodings.h:41:10: warning: multi-character character constant [-Wmultichar]
   41 |     m = 1'000'000ULL,
      |          ^~~~~

```

sipa commented:
@rebroad You need a C++17 compiler.
